src/V3Force.cpp

```diff
@@ -13,350 +13,297 @@
+            fc.m_enVscp->varp()->sigPublic(true);
+            fc.m_valVscp->varp()->sigPublic(true);
         }
     }
 
-    void changeRecurse(AstNode* nodep) {
-        // Recurse and replace any VarRef WRITEs to refer to the force equation
-        if (VL_LIKELY(!nodep->user1())) {  // Else processed already
-            if (auto* const varrefp = VN_CAST(nodep, NodeVarRef)) {
-                visitVarRef(varrefp);
-                return;  // Might have been edited -- and has no children so ok to exit
+    // CONSTRUCTOR
+    explicit ForceConvertVisitor(AstNetlist* nodep) {
+        // Transform all force and release statements
+        iterateAndNextNull(nodep->modulesp());
+
+        // Replace references to forced signals
+        nodep->modulesp()->foreach<AstVarRef>([this](AstVarRef* nodep) {
+            if (ForceComponentsVarScope* const fcp
+                = m_forceComponentsVarScope.tryGet(nodep->varScopep())) {
+                switch (nodep->access()) {
+                case VAccess::READ:
+                    // Read references replaced to read the new, possibly forced signal
+                    if (!nodep->user2()) {
+                        nodep->varp(fcp->m_rdVscp->varp());
+                        nodep->varScopep(fcp->m_rdVscp);
+                    }
+                    break;
+                case VAccess::WRITE:
+                    // Write references use the original signal
+                    break;
+                default:
+                    nodep->v3error(
+                        "Unsupported: Signals used via read-write reference cannot be forced");
+                    break;
+                }
             }
-        }
-        nodep->prefetch();
-        if (nodep->op1p()) changeRecurse(nodep->op1p());
-        if (nodep->op2p()) changeRecurse(nodep->op2p());
-        if (nodep->op3p()) changeRecurse(nodep->op3p());
-        if (nodep->op4p()) changeRecurse(nodep->op4p());
-        if (nodep->nextp()) changeRecurse(nodep->nextp());
+        });
     }
 
-    virtual void visit(AstNode* nodep) override { v3error("Unused"); }  // LCOV_EXCL_LINE
-
 public:
-    // CONSTRUCTORS
-    explicit ForceReplace(AstNetlist* nodep) { changeRecurse(nodep); }
-    virtual ~ForceReplace() override = default;
+    static void apply(AstNetlist* nodep) { ForceConvertVisitor{nodep}; }
 };
 
 //######################################################################
-// Force class functions
+//
 
 void V3Force::forceAll(AstNetlist* nodep) {
     UINFO(2, __FUNCTION__ << ": " << endl);
-    {  // Destruct before final check
-        ForceVisitor visitor{nodep};
-        if (visitor.anyForce()) {
-            V3Global::dumpCheckGlobalTree("force-mid", 0,
-                                          v3Global.opt.dumpTreeLevel(__FILE__) >= 3);
-            ForceReplace{nodep};
-        }
-    }
+    if (!v3Global.hasForceableSignals()) return;
+    ForceConvertVisitor::apply(nodep);
```

Comment:
If you want to propose we use this convention everywhere instead of calling the constructor we can, but until then I don't think it's helpful to do it differently in just this one Visitor.

Reply:
Yeah, TBH using "construct an object for it's side effects, but it must be in a local scope so the destructor is also invoked otherwise it leaves the word in a broken state" as a public API makes me twitch a bit. Also, visitors implement "algorithms", so semantically they are "invoked" rather than "instantiated", hence prefer a private constructor and the static entry point function (which also handles return values nicely, rather than having to poke around inside the implementation object).

FWIW I can remove it but I already sneaked in quite a couple of these :P I'm starting to see the pattern though, of you not liking these as incremental changes.

Reply:
I don't like incremental changes that are could be global but land in one visitor when it's added. If that was the general rule, every single visitor file would by now look quite different, and it would be much harder to read/understand them.  Thus when there's a new good "incremental idea", I like them to be applied to the entire code base.  The end result is the whole code base might look different from an older rev, but it's in a better and more consistent whole state.

Reply:
That's all fair, that methodology is just a bit problematic if the priority of "get this feature in and using this nice new abstraction I just came up with would make that easier" is higher than the priority of "apply the nice new abstraction everywhere it applies". I agree, we should not add one-offs in every patch, but sometimes having to go back and change everything first or worse get the feature in with less clear abstractions because it's urgent then forget about the improvement is a bit counter productive.

Would you be happy to add these where needed but also file an issue to either apply it everywhere or revert if turns out not to be generally applicable?

Reply:
>Would you be happy to add these where needed but also file an issue to either apply it everywhere or revert if turns out not to be generally applicable?

Ok.